feat: add critical TYPO3 patch level versions

madaxel opened the pull request:
Load available TYPO3 major versions and latest security version via
 TYPO3 API using CoreVersionService
This way we can add all TYPO3 core patch versions with known security issues as 'typo3-version.critical' to the output, allowing nagios to inform us about insecure installations

Hi @schams-net ,
related to https://github.com/schams-net/nagios_extensionlist/issues/12 my pull request as we tried to update the extension to TYPO3 v11. The patch worked for us the last 2 years. Maybe the PR can bring the function into your next release.
Tell me if I can help you

Regards Markus

mgerdes-atl committed:
feat: add critical TYPO3 patch level versions

* refactor internal marked CoreVersionService including CoreVersion
 classes from core into nagios_extensionlist
* pass over complete MajorRelease object to fluid instead of insecure
core versions only

schams-net commented:
Hi Markus. I like the idea and I'd add this feature to the extension. However, while reviewing your suggestion, the following issues came to my mind:

- The core class `\TYPO3\CMS\Install\Service\CoreVersionService\CoreVersionService` is marked as `@internal` and must not be used by extensions. We need an alternative approach or have to replicate its functionality.
- As the additional requests to the TYPO3 API take some time, we should cache the results for x hours by using TYPO3's [caching framework](https://docs.typo3.org/m/typo3/reference-coreapi/11.5/en-us/ApiOverview/CachingFramework/Index.html).
- I wonder if we should iterate the TYPO3 Core versions returned by the TYPO3 API and filter invalid versions such as "8.7-snapshot-20170404".
- We should implement some error handling if requests to the TYPO3 API fail (untested).
- I'd like to change the method names in the class `ExtensionlistController` as (for example) the method `listInsecureExtensionsAction()` does not only list extensions anymore 🙂
- ...

I started a new Git branch [issue/12](https://github.com/schams-net/nagios_extensionlist/tree/issue/12) to work on these changes.

madaxel commented:
Hi @schams-net ,
I refactored lots of stuff to be more reusable and use all major releases which are populated by the TYPO3 API
* removed the internal marked TYPO3CoreService
* First used CoreRelease and MajorRelease classes of typo3/cms-install but then created own objects to get more informations stored
* Iterateing over Releases of the API now, but not filtering, yet


I'm not familiar with the caching Framework so not a fast task for me ;-(
If we use a cache maybe we can store the resulting cfg somehow as backup. Or we could just write the file into a know tempfile which can be resend upon API Errors.

Done for today ;-)

schams-net commented:
Hi @madaxel. I merge your changes into branch #12 (commit ID 795971505f4a0a5cd1316b295f510999647a862a), made some adjustments, and implemented caching of the API responses by using TYPO3's Cache Framework (commit ID ae17443efe05cb83063aefedd73a25003109303f).

Still work-in-progress but we are getting closer 😎
